### ISSUE TITLE
Champion Allow Generic Attributes

### ISSUE THREAD

**Issue by gafter** [dotnet/csharplang#124]
- [ ] Proposal added
- [ ] Discussed in LDM
- [ ] Decision in LDM
- [ ] Finalized (done, rejected, inactive)
- [ ] Spec'ed

See
- https://github.com/dotnet/roslyn/issues/953
- https://roslyn.codeplex.com/discussions/542178
- https://github.com/dotnet/roslyn/pull/16814
- https://github.com/dotnet/coreclr/pull/9189

**Comment by AlphaDriveLegal** [comment]
I've wanted this for a long time!

**Comment by Mafii** [comment]
I would be happpy to have this, too!

**Comment by weitzhandler** [comment]
Very compelling.

Here's my use case, one among many:

    public class ValidateIf<TValidationAttribute> : ValidationAttribute
    {
      public TValidationAttribute Attribute { get; private set; }
      //other properties to resemble condition
      public ValidateIf(Action<TValidationAttribute> configure)
      {
        Attribute = configure();
      }
      protected override ValidationResult IsValid(object value, ValidationContext validationContext)
      {
        var shouldValidate = this.ShouldValidate();
        if(!shouldValidate) return ValidationResult.Success;
        else return Attribute.GetValidationResult(value, validationContext);
      }
    }

**Comment by Joe4evr** [comment]
@weitzhandler Your example is pretty flawed since lambdas can't be passed as an attribute parameter; only primitives, `string`, `typeof()` expressions, and `DateTime` literals (VB).

A more sensible example of where a generic attribute would be useful:
```cs
// This concerns a bot command system that can parse strings into typed arguments,
// but sometimes the "default" reader for something may be too narrow,
// so this on a parameter would override which class does the reading for that command.
[AttributeUsage(AttributeTargets.Parameter)]
public sealed class OverrideTypeReaderAttribute<TReader> : Attribute
    where TReader : TypeReader
{
    // We're only interested in the Type info at this point,
    // but we would really like the type-safety of the generic constraint
    public Type TypeReaderType { get; }

    public OverrideTypeReaderAttribute()
    {
        TypeReaderType = typeof(TReader);
    }
}
// Usage:
public Task Foo([OverrideTypeReader<AlternativeBoolReader>] bool switch)
{
    //.....
}
```
Currently we imitate the constraint by taking a `Type` parameter in the constructor and checking `IsAssignableFrom()` at runtime, but pushing the check to compile-time via constraint would be neat and save us from needing to throw an exception if it fails.

**Comment by weitzhandler** [comment]
@Joe4evr 
I thought part of this proposal was allowing complex params I attribute initialization, which I saw in several issues, maybe in the Roslyn repo.
And that's why I was going for an example that utilizes both, and I do vote for both genetic attributes and better and more flexible attribute initialization and setup options.

**Comment by Joe4evr** [comment]
> I thought part of this proposal was allowing complex params I attribute initialization

I have not seen this proposal before, not to mention that such a thing would require quite a hefty spec change since as it stands, the compiler needs to be able to serialize the arguments directly in the metadata.

It *is* possible, if wanted for whatever reason, to instantiate an attribute as just a normal class, and that could use any type of parameter:
```cs
var attr = new SomethingSpecialAttribute(GetSomeData());
```
But that should only be a secondary constructor, otherwise it's totally useless to have that class be an attribute in the first place. (I have an example of where this is used in the same codebase I described above, but that's getting off-topic.)

**Comment by gafter** [comment]
> I thought part of this proposal was allowing complex params I attribute initialization,

No.

**Comment by alrz** [comment]
>We need a mechanism to understand which target runtime it works on. We need that for many things, and are currently looking at that. Until then, we can't take it.

I suppose that's on the use-site, whereas an abstract generic attribute doesn't seem to be harmful,
```cs
abstract class SomeAttribute<T> : Attritbue {}
```
Could we permit the above without (any) runtime support?

**Comment by ymassad** [comment]
Would this feature support this?

```
public class GenericClass<T>
{
    [SomeAttribute<T>]
    public void DoSomething(T input)
    {
    }
}

public class SomeAttribute<T> : Attribute
{

}
```

**Comment by ymassad** [comment]
@roydukkey , I am not sure I understand. Can you elaborate?

Is B a concrete type?

**Comment by roydukkey** [comment]
@ymassad Sorry. I meant to write this.

```csharp
[SomeAttribute<B>]
public void DoSomething<B>(B input) { }
```

I believe this is the spec.

**Comment by AviAvni** [comment]
The attribute can't have open generic type it must be instantiated at compile time

**Comment by jnm2** [comment]
@ymassad What attribute instance would you expect `typeof(GenericClass<>).GetMethod("DoSomething").GetCustomAttributes()` to contain in your example?

**Comment by ymassad** [comment]
@jnm2 ,

Thanks. I see now why this is not possible.

what about this:

```
public class GenericClass<T>
{
    [SomeAttribute(typeof(T)]
    public void DoSomething(T input)
    {
    }
}

public class SomeAttribute : Attribute
{
    public SomeAttribute(Type type)
    {
        this.Type = type;
    }

    public Type Type {get;}
}
```

`typeof(GenericClass<>).GetMethod("DoSomething").GetCustomAttributes()` could return a `SomeAttribute` instance with `Type` containing the open type `T`.

**Comment by jnm2** [comment]
@ymassad The problem there (if I remember the metadata format correctly) is that attribute type arguments are stored as serialized strings containing the assembly-qualified name. What string can be deserialized similar to `Type.GetType` in such a way that it resolves to a generic type parameter? I think this would require a runtime change.

**Comment by jnm2** [comment]
Yep, https://www.ecma-international.org/publications/files/ECMA-ST/ECMA-335.pdf, section **II.23.3 Custom attributes**, page 268:

> - If the parameter kind is _System.Type_, (also, the middle line in above diagram) its value is stored as a 
_SerString_ (as defined in the previous paragraph), representing its canonical name.  The canonical name is its full type name, followed optionally by the assembly where it is defined, its version, culture and public -key-token.  If the assembly name is omitted, the CLI looks first in the current assembly, and then in the system library (mscorlib); in these two special cases, it is permitted to omit the assembly-name, version, culture and public-key-token.

**Comment by ymassad** [comment]
@jnm2 , thanks.

**Comment by jnm2** [comment]
@ymassad TypeDescriptor will inject the type as a constructor parameter when it instantiates a type converter specified via TypeConverterAttribute, and maybe you can follow a similar pattern in the meantime?

https://github.com/dotnet/corefx/blob/bfb4ba1e398273cfd3bf15bf25c535f64bf072ea/src/System.ComponentModel.TypeConverter/src/System/ComponentModel/MarshalByValueComponent.cs#L16

https://github.com/dotnet/corefx/blob/bfb4ba1e398273cfd3bf15bf25c535f64bf072ea/src/System.ComponentModel.TypeConverter/src/System/ComponentModel/ComponentConverter.cs#L13-L18

https://github.com/dotnet/corefx/blob/bfb4ba1e398273cfd3bf15bf25c535f64bf072ea/src/System.ComponentModel.TypeConverter/src/System/ComponentModel/ReflectTypeDescriptionProvider.cs#L223-L232

**Comment by rhyous** [comment]
The main use case I have is to make the attribute know the type I am placing it on.

```
public class EntityExampleAttribute<T> : Attribute
{
    public abstract List<T> Examples {get;set;}
}
```
Or
```
public class EntitySeedDataAttribute<T> : Attribute
{
    public abstract List<T> Entities {get;set;}
}
```
Then I can do inheritance.

```
public class Entity1Attribute : EntityExampleAttribute<Entity1>
{
    public Entity1Attribute() 
    {
        Examples = new Entity1 { /*   example entity */ }
    }
}
```

**Comment by dzmitry-lahoda** [comment]
Will it be part of C# 8.X release? Can I start using it with some preview version?

Sorry for by broken C#, as it is not my native language. But I hope to write next kind of DSL with Roslyn generator
```csharp
// union which stores some stucture so that no need to box with explict layout
[OverlappedUnion(Size = 64)] // analyar will thorw if any size is more than 64
//[OverlappedUnion(Size = 64, ReadOnly = true)]
enum X
{
  [Case<String>]
  A,
    [Case<(String z, int k>>] // x.B.zm x.B.k
    B,
    
    // no data
    [Case]
    [Case<Unit>]
    C,
    
    // fixed(see C# keyword) allocated 23 length char
    [Case<Memory<char>>(23)]
    E,
    
    [Case("CustomPropertyName")]
    D,
}

// generator generates all possible cases
XCase x;

switch (x.Tag)
{
   case X.A:
     // in DEBUG runtime checks A and then does not thorows A
     // in analyzer also checks
     // should work well with descruting
      var v1 = x.A;
     var v2 = x.Value; // object of Union or XStruct for OverlappedUnion
    break;
   case X.C
     break;
   // analyzer ensure default or all cases
}


// design with internal explicit stuct or base class or what ever,
// same API but may require boxing
[Union]
enum X
{
  [Case<String>]
  A,
    [Case<(String z, int k>>]
    B,
    
    // no data
    [Case]
    [Case<Unit>]
    C,
        
    [Case<Memory<char>>(23)]
    E,    
}
```

**Comment by Joe4evr** [comment]
> Will it be part of C# 8.X release? Can I start using it with some preview version?

Sadly, [despite the Roslyn PR being ready for literally a year now](https://github.com/dotnet/roslyn/pull/26337), it doesn't seem like it's gonna make 8.0 unless the language team decides to push the Merge button at literally the last minute.

**Comment by thiagomajesk** [comment]
@Joe4evr Just came across this today and I have to say that this would indeed be awesome to have right away with C# 8. BTW, it seems that something is going int this PR since your comment. 
Do you guys (@jcouv @AviAvni) mind bringing some information over here about this feature's availability?

**Comment by ufcpp** [comment]
https://github.com/dotnet/roslyn/blob/master/docs/Language%20Feature%20Status.md

![image](https://user-images.githubusercontent.com/3760955/58849715-648b2d00-86c6-11e9-9ed7-667afca85dc0.png)

**Comment by jnm2** [comment]
> @ymassad The problem there (if I remember the metadata format correctly) is that attribute arguments of type System.Type are stored as serialized strings containing the assembly-qualified name. What string can be deserialized (similar to `Type.GetType`) in such a way that it resolves to a generic type parameter? I think this would require a runtime change.

I reread this (https://github.com/dotnet/csharplang/issues/124#issuecomment-444965617) and have to correct myself:
`!0` (referring to the generic type parameter with index 0) and `!!0` (same but for generic method parameters) are how this would work.

**Comment by TonyValenti** [comment]
Any chance at all this will make it into C# 9?

**Comment by Joe4evr** [comment]
@TonyValenti Given that this issue is labeled as "10.0 Candidate", the answer is "No".

**Comment by YairHalberstadt** [comment]
Another use case for this:

I'm writing a compile time IOC framework for C# using source generators: https://github.com/YairHalberstadt/stronginject

Registrations are done via attributes, but because generics aren't allowed in attributes you can't do this: `[RegisterGeneric(typeof(List<>))]`

Instead you have to explicitly pass in the parameters using a generic factory method, e.g:

```csharp
[Factory] public static List<T> CreateList<T>(IEnumerable<T> ts) => new List<T>(ts);
```

Whilst this is more flexible it's less concise, and requires you to update the method whenever the parameters change, which is against the spirit of using an IOC container in the first place.

Another thing that might be useful is partially open typeofs, allowing you to register e.g. `[RegisterGeneric(typeof(ServiceImpl<string, >), typeof(IService<>))]`.

**Comment by johnscott999** [comment]
While the use cases for this are not broad, it would have a big impact on certain applications.

For example, MVC's `ProducesResponseTypeAttribute(Type type, int statusCode)` can specify different result types based on the HTTP status. Currently, the following code does not build:
```csharp
    internal abstract class BaseCrudController<TEntity> : Controller
        where TEntity: class
    {
        private DbContext context;

        //...

        // This next line throws an error
        [ProducesResponseType(typeof(TEntity), 200)]
        //                    ^^^^^^^^^^^^^^^ 
        //                    CS0416: An attribute argument cannot use type parameters
        [ProducesResponseType(400)]
        public virtual async Task<IActionResult> GetById(int id)
        {
            var instance = await this.context.Set<TEntity>().FindAsync(id);
            return instance is null 
                ? (IActionResult)NotFound()
                : (IActionResult)Ok(instance);
        }

        //...

    }
```
With this feature the annotation could be rewritten as
```csharp
        [ProducesResponseType<TEntity>(200)]
        [ProducesResponseType(400)]
        public virtual Task<IActionResult> GetById(int id)
        {
            //...
```
Inheritors of this base could specify a closed generic, if compile time typing was needed, like for a swagger doc or api code gen.
```csharp
    public class UserController: BaseCrudController<User>
    {
        //...
```
I would expect reflection calls returning that attribute to behave similarly to reflection calls on generic properties: If the generic is defined in the originating type definition, then return the closed generic. If the originating type is open, return the open definition.

![image](https://user-images.githubusercontent.com/47427491/100008010-7f59d180-2d92-11eb-9daa-411cf79a98df.png)

**Comment by julealgon** [comment]
@johnscott999 you can achieve identical behavior by leveraging `ActionResult<T>` instead of `IActionResult`. There is no need to explicitly specify a `ProducesResponseType` in that case.

Having said that, I can see your suggestion being very useful in other contexts.

**Comment by mrtristan** [comment]
@julealgon not so if you have many instances of `ProducesResponseType` such as for cases where a bad request responds with a different model.

**Comment by julealgon** [comment]
> @julealgon not so if you have many instances of ProducesResponseType such as for cases where a bad request responds with a different model.

I was specifically talking about the example provided, where it was being used with the entity type and the 200 response (a common thing). `ActionResult<T>` solves that particular issue without needing generic types in attributes.

**Comment by AlFasGD** [comment]
dotnet/roslyn#16814 was closed, the new PR is dotnet/roslyn#26337

**Comment by hez2010** [comment]
@johnscott999 Even with generic attributes feature you still cannot write code like that. The runtime doesn't allow to use open generics in attributes.

**Comment by johnscott999** [comment]
@hez2010 the last code sample was my expectation on how reflection calls would interact with generic attributes. I'd expect generic attributes to be defined as either a defined type, or, if decorating an generic class or it's properties, one of that class's generic arguments.

```csharp
//Valid
[GenericClassAttr<T>]
public class GenericClass<T> {
    [GenericPropAttr<T>]
    public string Property { get; set; }
    [GenericPropAttr<string>]
    public string Property { get; set; }
    // ...
}
    
// Invalid
[GenericClassAttr<TAnother>]
public class GenericClass<TOne> {
    // ...
}
    
// Invalid
public class GenericClass<TOne> {
    [GenericPropAttr<TAnother>]
    public string Property { get; set; }
    // ...
}
```

@julealgon You can work around it, but that also means you can't, for example, use the controller methods for non-success codes `return NotFound()` if you have a bad id on a get. MVC is flexible enough to work with and around, but not all use cases are as versatile.

**Comment by julealgon** [comment]
@johnscott999 
> that also means you can't, for example, use the controller methods for non-success codes return NotFound() if you have a bad id on a get. 

Why not? All action results in AspNetCore can be converted implicitly to `ActionResult<T>`. Please try it yourself. You can _definitely_  have a `ActionResult<MyEntity>` returning controller action and do a `return NotFound()` from it.

Of course, you have to then explicitly add the `[ProducesResponseType]` for those other cases too (as one would expect).

**Comment by johnscott999** [comment]
@julealgon Huh, I'd always assumed that `IActionResult<T>` ineherited from `IActionResult`, but that's not it at all. You learn something every day. Regardless, this isn't the an aspnetcore issue, I'm was just trying to use that library to illustrate a use case. 

Ultimately my argument is that having generic types in attributes would be helpful when using a abstract base class with inheritance that has attribute decorations. And since you would need to add the produces response type, having a crud controller for mvc webapi is still demonstrative, if not idea.

**Comment by RikkiGibson** [comment]
One thing that I haven't seen discussed in this issue is how `AllowMultiple = false` should be handled.

```cs
using System;

[AttributeUsage(AttributeTargets.All, AllowMultiple = false)]
class Attr<T> : Attribute { }

[Attr<string>]
[Attr<string>] // error
[Attr<object>] // error?
class Program
{
}
```

The question is: if `[Attr<string>]` and `[Attr<object>]` are both applied to something, should it be considered a violation of the `AllowMultiple` setting?

Tentatively I am inclined to go with the most restrictive possible interpretation, and make it so different constructions of the same attribute is an error with `AllowMultiple = false`. [Back in 2015](https://github.com/dotnet/roslyn/issues/953#issuecomment-77230025) this was suggested (emphasis mine):

> > How does  AttributeUsage.AllowMultiple = false  work with generic attributes: does it limit the usages to one and only one attribute instance, or one instance per generic type?
> 
> Probably it doesn't because the rules for this case haven't been written. `FooAttribute<int>` and `FooAttribute<string>` are different types so they would probably be both allowed without a special rule. **The rule should be that the generic definition - `FooAttribute<>` - is considered rather than its instantiations.**

**Comment by TonyValenti** [comment]
@RikkiGibson -
I think that in the scenario you mentioned, if you want to prohibit MyAttribute<A> and MyAttribute<B>, it would be best to introduce a base MyAttribute (which both generics inherit from) class and apply the AllowMultiple(false) attribute to it.

**Comment by RikkiGibson** [comment]
I don't expect that to introduce the expected restriction because the following currently compiles just fine: [SharpLab](https://sharplab.io/#v2:EYLgtghglgdgNAFxFANgHwAICYCMBYAKEIG0BBBBAJymAFcEBTAVQGcIBzBgCnKpvoYAVCJU4IWAOlIoUcAATSUAewDuAWVooEUAA4oGcgLxyAZhBQsGASgC6hbAoqU5IR3zqM5AbzkBfQvZYbpQ4LsHefoHBQa68zj7+RARkTjh2yXFY6RgAzHIOAMLehHKl+XkYACxyalxWxQRlkQSJQA=)

```cs
using System;

[AttributeUsage(AttributeTargets.All, AllowMultiple = false)]
class Attr : Attribute { }

class Attr1 : Attr { }
class Attr2 : Attr { }

[Attr1]
[Attr2]
public class C {
    public void M() {
    }
}
```

**Comment by AlFasGD** [comment]
My best idea is to have AllowMultiple only prevent using multiple of the same attribute including the type arguments. For example, `Attr<string>` and `Attr<object>` will be usable with the AllowMultiple property set to true.

What I would instead use is a new property in `AttributeUsage` that determines whether to prohibit the same attribute with any type arguments, which I'd call `AllowMultipleGenericUnbound` (there surely is a better name for that). Defaults to `true`, when `true` it allows using that `Attr` with the same type arguments multiple times, and when `false`, it prohibits using the attribute more than once regardless of its type arguments. It obviously has no effect on non-generic attributes.

**Comment by HaloFour** [comment]
I'm curious as to if/how the runtime enforces that restriction today given IIRC this was already shipped in the runtime.

**Comment by jnm2** [comment]
This feels very similar to the convo at https://github.com/dotnet/roslyn-analyzers/issues/5225#issuecomment-877242522.

**Comment by Shadowblitz16** [comment]
I like this idea but it only solves part of my problem

```

[RequireComponent<Transform, Renderer, Collider>()]
public class MyEntity : Entity
{

}

// ??? params on generics isn't supported
```

**Comment by HaloFour** [comment]
@Shadowblitz16 

> ??? params on generics isn't supported

The runtime has no concept of [variable generic arguments](https://github.com/dotnet/roslyn/issues/5058).  You would need to define multiple versions of that attribute for each possible arity of generic arguments.  Or you could just use a `params Type[]` constructor as you aren't really taking advantage of generics here except to have a different syntax:

```cs
[RequireComponent(typeof(Transform), typeof(Renderer), typeof(Collider))]
public class MyEntity : Entity
{

}
```

**Comment by Shadowblitz16** [comment]
@HaloFour the issue is disallowing  anything that is not a component to be a compile time error.

**Comment by HaloFour** [comment]
@Shadowblitz16 

> the issue is disallowing anything that is not a component to be a compile time error.

Right, if you want to use generic type constraints then you'll need to define multiple versions of that attribute, one for each number of generic type arguments you can accept:

[SharpLab](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQHMIDsIIJYDGqsM+wcMKANDCAG655gCeAPgAIBMAjALABQA9gGYABHiyUEYKAQiiAkgGEA9gFsADipyTRAb1EBfAULESpMuYoBKEAI5w8SACarN27DGT6Bov6IAVZg0IAG0AXVFbBycIV3UtHS99UUwYAG4jAWNBfhFRLij7Rxc3RM9kAEEYUjxySgAeAIA+URBRatr6iCobYtj49yTkX38AdwALXHkAtsUyj0kBPVG/IJCIvpjShMWvADpokriF4dEAXlacMdEIgxhgiBUwAAoAgEojdOyTPLFCo4DU4VTpkCgQJrcXoBTitdqgurg3oKQE7IYVVaiSbTQLcObKXZJTHYpCBTj44FLfgrfj+QKPTYo/po8qSZCHZknQkVC5XCA3O6iB4hZ5vbjvXrCp6vGGfQzffg5UwFcmornotkI7qQ6GcaHCOEdGpgyjItWDVleYlTUkBPHtAkamDWnEwincql0kkzMQOynO6mY9ZhSJM7bqy3s83+7yXUTXW7he6PUV2iVClMyzjpqWp4RyhVKv4qorhi17KrGxGNO26/XQgAshq1SK2x3Lwxdtvt8w9Aa9Npm5L9fa7PvdTrHgQbE8tyyDDND0b7Uc5HZ5cYTgtzMvFkszb2z+5FMvzx+lbwbBZ+uXyhX9PcdlpSOTv5P9w97Tpfyvvfd9X7PgYRYCKEy5OsgDQPr0H7NOEv7vj4tL+GBa4xlBfZQqIH4wf+cGYneM4ALIvJ8NJ0jkhhAA)

One can argue that variadic generics would make your use case easier, but it gets substantially more complicated for any use case that would want to use those generic type arguments for anything other than reflection purposes.

**Comment by Shadowblitz16** [comment]
@HaloFour this would work very well with variadic generic paramaters

**Comment by RikkiGibson** [comment]
dotnet/roslyn#55577

We will make this feature available in the "preview" language version in .NET 6 RC1. We found that a large number of runtimes and tools (Mono, C++/CLI, etc.) may still have issues consuming generic attributes, so we are going to take an additional release cycle to try and iron out the issues, and give ourselves some time to make breaking changes with runtime APIs.

For example, if a user calls `CustomAttributeData.GetCustomAttributes(someType, typeof(GenericAttr<>))` there is an open question about whether it should give all the constructions of the attribute which were used on the type. dotnet/runtime#56887

```cs
using System;

// what should the value of 'attrs' be here?
var attrs = CustomAttributeData.GetCustomAttributes(typeof(C), typeof(Attr<>));

[AttributeUsage(AttributeTargets.All, AllowMultiple = true)]
class Attr<T> : Attribute { }

[Attr<int>]
[Attr<string>]
class C { }
```

**Comment by johnscott999** [comment]
It’s a bit cheeky, but why not `Attribute[]`

```csharp
Attribute[] attrs = //…
attrs[0].GetType() // => typeof(Attr<int>)
```

from there you can get what you need with reflection.

practically I’d probably, inherit from a non generic base type that exposed the generic types of `derived` as properties, but that’s a design decision. 

as for allow multiple, I think you’ve got to treat each different generic as a different class, the same way you would static properties in a generic class. I’d support adding an attribute usage for `AllowMultipleGeneric` (name pending), that would only allow one `Attr<T>` decorator if set. That would cover all the use cases that come to mind

**Comment by Joe4evr** [comment]
> practically I’d probably, inherit from a non generic base type that exposed the generic types of `derived` as properties

I have a feeling that's going to be the majority case anyway.

> `// what should the value of 'attrs' be here?`

Looking at the [existing overloads for `CustomAttributeData.GetCustomAttributes`](https://apisof.net/catalog/b358fbba-ad3a-4f30-40a4-505ee037964c), they all return an `IList<CustomAttributeData>` so I see no reason to deviate from that. If called with an open/half-open type, I'd expect it to contain the data of all the attributes applied with the "missing" type arguments filled in to what they are on the use-site, or an empty collection if none matched.
```cs
[AttributeUsage(AttributeTargets.All, AllowMultiple = true)]
class Attr<T1, T2> : Attribute { }

[Attr<long, int>]
[Attr<long, string>]
[Attr<int, string>]
[Attr<object, string>]
class C { }

// { Attr<long, int>, Attr<long, string>, Attr<int, string>, Attr<object, string> }
var attrs0 = CustomAttributeData.GetCustomAttributes(typeof(C), typeof(Attr<,>));

// { Attr<long, int>, Attr<long, string> }
var attrs1 = CustomAttributeData.GetCustomAttributes(typeof(C), typeof(Attr<long,>));

// { Attr<long, string>, Attr<int, string>, Attr<object, string> }
var attrs2 = CustomAttributeData.GetCustomAttributes(typeof(C), typeof(Attr<,string>));

// { } (none of the applied attributes have 'object' as T2)
var attrs3 = CustomAttributeData.GetCustomAttributes(typeof(C), typeof(Attr<,object>));
```
Does that make sense?

I think it may be beneficial to add a new `TypeArguments` property to `CustomAttributeData` as well, even though they can also simply be fished out of `AttributeType`, it could be more discoverable for people already working with this a lot.

**Comment by AlFasGD** [comment]
The catch here is that you cannot partially bind a generic type in a `typeof` expression; and I can't remember if you're allowed to not fully bind an open generic type via reflection. I can see why partial generic binding shouldn't be prohibited, and I support it.

> as for allow multiple, I think you’ve got to treat each different generic as a different class, the same way you would static properties in a generic class. I’d support adding an attribute usage for `AllowMultipleGeneric` (name pending), that would only allow one `Attr<T>` decorator if set. That would cover all the use cases that come to mind

https://github.com/dotnet/csharplang/issues/124#issuecomment-880150312

---

Now, onto the real stuff:

- Generic types are immediately ignored on certain implementations of `GetCustomAttributes`. That's because up until now, attributes couldn't be generic. This would have to change accordingly after generic attributes go live.
- Statically typed open generic types aren't usable outside of `typeof` contexts, meaning you'd have to bind the type arguments of a generic attribute instance.

Those arguments are enough to show that an open generic attribute type can function as a wildcard to return all attributes that are of the type.

The other option is to discard that option and throw an exception. I see no reason to invalidate functionality that cannot be otherwise utilized. It should however be properly marked down in the documentation that open generic attribute types result in all attribute instances binding to the specified generic type.

**Comment by TheFanatr** [comment]
I totally +1 the idea of allowing partial generic binding to be a thing, but why stop at `typeof`? If static generic type reference syntax needs to be improved to support constraining to specific parameters, via relaxing rules around generic parameters, I'd say that this should be done more broadly and not focus on just this specific use of generic syntax.

> Statically typed open generic types aren't usable outside of typeof contexts, meaning you'd have to bind the type arguments of a generic attribute instance.

@AlFasGD Is there some technical reason behind this (beyond the obvious)?

I understand that in most cases the generic type would need to be interacted with (so not talking about trivial use of `is`) and so would need to have all the types resolved to use the full type surface in non-trivial cases, but this can be worked around by not allowing access to members which expose undefined generic parameters and/or exposing the least derived compatible type based on the variance of the parameters.

More than this just being convenient, the `is` syntax was initially basically meant to replace and make `(...).GetType() == typeof(...)` more optimizable anyways, so if you can use open generics in `typeof`, then it would make sense to allow them, at the bare minimum, in the trivial `is` syntax, but I'd personally aim a little higher. To be clear, while I understand it is not possible to constrain the earlier-mentioned type check that uses `==` with the proposed changes to partial generics syntax, it would still be possible using the reflection APIs above in the thread, so it seems like a reasonable extension of the syntax sugar for `is`, additionally considering that `GetType` and `typeof` can be considered reflection, and that this is all still static (not dynamic) if variance-based inference is used as suggested to fill in the unspecified parameters and maybe with other solutions as well.

Let's take an example.

```csharp
Object table = new Dictionary<string, bool> { ["e"] = true };

if (table is Dictionary<string,> referenceTable) {
    Console.WriteLine(referenceTable.ContainsKey("e"));
}

if (table is Dictionary<,bool> referenceTable) {
    Console.WriteLine(referenceTable.Values[0]);
}
```

With that, it would seem reasonable to then also enable similar access to static members, although I recognize it is a little different.

```csharp
static class TestClassA<T> {
    string testA = "";
}

...

TestClassA<>.testA = ...;
```

And I mean, while I'm at it, may as well point out this one too. (Not open generic type, but still a partial generic)

```csharp
public static TNotInferrable TestA<TInferrableA, TInferrableB, TNotInferrable>(TInferrableA inferrableA, TInferrableB inferrableB) => ...;
public static void TestB<TInferrableA, TInferrableB>(TInferrableA inferrableA, TInferrableB inferrableB) => ...;

...

var success = TestB<,,bool>("", 341.56D);
TestA("", 341.56D); // currently legal
```

Basically if you're going to relax restrictions on generic syntax to allow constraining by specific parameters for the sake of convenience, may as well apply that wherever possible.

Just some thoughts. Sorry for the messy writing, late at night (early in the morning).

**Comment by AlFasGD** [comment]
That's an already open proposal, you can find it by looking for "unbound", "partially bound", "partially open" generics

**Comment by seanblue** [comment]
Is it possible for generic attributes to work on .NET Framework? In the past, some "unsupported" features have become compatible just by copying a few types (e.g. `init` keyword) and other times it's impossible (as far as I know). While I am able to define generic attribute types and apply them to classes/methods/etc., the methods for actually *finding* the attributes used (e.g. `MemberInfo.GetCustomAttributes`) throw an exception when passing in a generic attribute. Are there any types I can copy or otherwise anything I can do to fully enable this feature on .NET Framework?

**Comment by hez2010** [comment]
> Are there any types I can copy or otherwise anything I can do to fully enable this feature on .NET Framework?

Nope. The generic attributes feature requires a runtime change and the change won't be backported to .NET Framework.

**Comment by seanblue** [comment]
@hez2010 I had a feeling that would be the case. Thanks for confirming. 🙂

**Comment by BlinD-HuNTeR** [comment]
@seanblue It should be possible, however, to write something that directly reads and parses metadata from the assembly file, without using the reflection infrastructure. Probably something based off of Mono.Cecil would do the job, depending on what you want.

**Comment by TonyValenti** [comment]
In the past, to get around this, I will have a generic attribute base class and then create a non-generic attribute that inherits from it and use that.

I also use a similar approach when I want to provide values to an attribute that are classes/disallowed types.

ie: ```
public class MyAttribute<T> : Attribute { }
public class MyStringAttribute : MyAttribute<string> { } 

[MyStringAttribute]
public class Foo { }

```
